# How do I merge DataFrames in pandas? ([video](https://www.youtube.com/watch?v=iYWKfUOtGaw&list=PL5-da3qGB5ICCsgW1MxlZ0Hq8LL5U3u9y&index=32))

- [My pandas video series](https://www.dataschool.io/easier-data-analysis-with-pandas/)
- [GitHub repository](https://github.com/justmarkham/pandas-videos)

## Table of contents

1. Selecting a Function
2. Joining (Merging) DataFrames
3. What if...?
4. Four Types of Joins

# Part 1: Selecting a Function

Taken from [Merging DataFrames with pandas](https://www.datacamp.com/courses/merging-dataframes-with-pandas?tap_a=5644-dce66f&tap_s=280411-a25fc8) (DataCamp course):

- `df1.append(df2)`: stacking vertically
- `pd.concat([df1, df2])`:
  - stacking many horizontally or vertically
  - simple inner/outer joins on Indexes
- `df1.join(df2)`: inner/outer/left/right joins on Indexes
- `pd.merge(df1, df2)`: many joins on multiple columns

# Part 2: Joining (Merging) DataFrames

Using the [MovieLens 100k data](http://grouplens.org/datasets/movielens/), let's create two DataFrames:

- **movies**: shows information about movies, namely a unique **movie_id** and its **title**
- **ratings**: shows the **rating** that a particular **user_id** gave to a particular **movie_id** at a particular **timestamp**

In [1]:
import pandas as pd
from ds_utils.ds_preamble import *
from ds_utils.ds_plotting import *
from ds_utils.ds_helper import *
import datetime
import duckdb
pd.options.display.max_columns = 100

%load_ext autoreload
%autoreload 2

calling ds_preamble..
calling ds_helper...


In [51]:
from tabulate import tabulate
# testings()
# tips = sns.load_dataset('tips')

# print(tips[:30].to_markdown())

# print(tips[:30])


|    |   total_bill |   tip | sex    | smoker   | day   | time   |   size |
|---:|-------------:|------:|:-------|:---------|:------|:-------|-------:|
|  0 |        16.99 |  1.01 | Female | No       | Sun   | Dinner |      2 |
|  1 |        10.34 |  1.66 | Male   | No       | Sun   | Dinner |      3 |
|  2 |        21.01 |  3.5  | Male   | No       | Sun   | Dinner |      3 |
|  3 |        23.68 |  3.31 | Male   | No       | Sun   | Dinner |      2 |
|  4 |        24.59 |  3.61 | Female | No       | Sun   | Dinner |      4 |
|  5 |        25.29 |  4.71 | Male   | No       | Sun   | Dinner |      4 |
|  6 |         8.77 |  2    | Male   | No       | Sun   | Dinner |      2 |
|  7 |        26.88 |  3.12 | Male   | No       | Sun   | Dinner |      4 |
|  8 |        15.04 |  1.96 | Male   | No       | Sun   | Dinner |      2 |
|  9 |        14.78 |  3.23 | Male   | No       | Sun   | Dinner |      2 |
| 10 |        10.27 |  1.71 | Male   | No       | Sun   | Dinner |      2 |
| 11 |      

In [45]:
testings()

TypeError: 'NoneType' object is not subscriptable

In [49]:
customers = read_from_db('SELECT * FROM customers c', db_name='book_db')
orders = read_from_db('SELECT * FROM orders o', db_name='book_db')
customers.merge(orders, how='left', on='customerid')
customers.merge(orders, how='inner', on='customerid')

customerid          companyname              contactname  \
0        ALFKI  Alfreds Futterkiste             Maria Anders   
1        ALFKI  Alfreds Futterkiste             Maria Anders   
2        ALFKI  Alfreds Futterkiste             Maria Anders   
3        ALFKI  Alfreds Futterkiste             Maria Anders   
4        ALFKI  Alfreds Futterkiste             Maria Anders   
..         ...                  ...                      ...   
827      WOLZA       Wolski  Zajazd  Zbyszek Piestrzeniewicz   
828      WOLZA       Wolski  Zajazd  Zbyszek Piestrzeniewicz   
829      WOLZA       Wolski  Zajazd  Zbyszek Piestrzeniewicz   
830      WOLZA       Wolski  Zajazd  Zbyszek Piestrzeniewicz   
831      WOLZA       Wolski  Zajazd  Zbyszek Piestrzeniewicz   

             contacttitle          address      city region postalcode  \
0    Sales Representative    Obere Str. 57    Berlin   None      12209   
1    Sales Representative    Obere Str. 57    Berlin   None      12209   
2    Sales Representative    Obere Str. 57    Berlin   None      12209   
3    Sales Representative    Obere Str. 57    Berlin   None      12209   
4    Sales Representative    Obere Str. 57    Berlin   None      12209   
..                    ...              ...       ...    ...        ...   
827                 Owner  ul. Filtrowa 68  Warszawa   None     01-012   
828                 Owner  ul. Filtrowa 68  Warszawa   None     01-012   
829                 Owner  ul. Filtrowa 68  Warszawa   None     01-012   
830                 Owner  ul. Filtrowa 68  Warszawa   None     01-012   
831                 Owner  ul. Filtrowa 68  Warszawa   None     01-012   

     country          phone            fax  orderid  employeeid   orderdate  \
0    Germany    030-0074321    030-0076545  10643.0         6.0  1997-08-25   
1    Germany    030-0074321    030-0076545  10692.0         4.0  1997-10-03   
2    Germany    030-0074321    030-0076545  10702.0         4.0  1997-10-13   
3    Germany    030-0074321    030-0076545  10835.0         1.0  1998-01-15   
4    Germany    030-0074321    030-0076545  10952.0         1.0  1998-03-16   
..       ...            ...            ...      ...         ...         ...   
827   Poland  (26) 642-7012  (26) 642-7012  10792.0         1.0  1997-12-23   
828   Poland  (26) 642-7012  (26) 642-7012  10870.0         5.0  1998-02-04   
829   Poland  (26) 642-7012  (26) 642-7012  10906.0         4.0  1998-02-25   
830   Poland  (26) 642-7012  (26) 642-7012  10998.0         8.0  1998-04-03   
831   Poland  (26) 642-7012  (26) 642-7012  11044.0         4.0  1998-04-23   

    requireddate shippeddate  shipvia  freight              shipname  \
0     1997-09-22  1997-09-02      1.0    29.46   Alfreds Futterkiste   
1     1997-10-31  1997-10-13      2.0    61.02  Alfred's Futterkiste   
2     1997-11-24  1997-10-21      1.0    23.94  Alfred's Futterkiste   
3     1998-02-12  1998-01-21      3.0    69.53  Alfred's Futterkiste   
4     1998-04-27  1998-03-24      1.0    40.42  Alfred's Futterkiste   
..           ...         ...      ...      ...                   ...   
827   1998-01-20  1997-12-31      3.0    23.79         Wolski Zajazd   
828   1998-03-04  1998-02-13      3.0    12.04         Wolski Zajazd   
829   1998-03-11  1998-03-03      3.0    26.29         Wolski Zajazd   
830   1998-04-17  1998-04-17      2.0    20.31         Wolski Zajazd   
831   1998-05-21  1998-05-01      1.0     8.72         Wolski Zajazd   

         shipaddress  shipcity shipregion shippostalcode shipcountry  
0      Obere Str. 57    Berlin       None          12209     Germany  
1      Obere Str. 57    Berlin       None          12209     Germany  
2      Obere Str. 57    Berlin       None          12209     Germany  
3      Obere Str. 57    Berlin       None          12209     Germany  
4      Obere Str. 57    Berlin       None          12209     Germany  
..               ...       ...        ...            ...         ...  
827  ul. Filtrowa 68  Warszawa     

customerid          companyname              contactname  \
0        ALFKI  Alfreds Futterkiste             Maria Anders   
1        ALFKI  Alfreds Futterkiste             Maria Anders   
2        ALFKI  Alfreds Futterkiste             Maria Anders   
3        ALFKI  Alfreds Futterkiste             Maria Anders   
4        ALFKI  Alfreds Futterkiste             Maria Anders   
..         ...                  ...                      ...   
825      WOLZA       Wolski  Zajazd  Zbyszek Piestrzeniewicz   
826      WOLZA       Wolski  Zajazd  Zbyszek Piestrzeniewicz   
827      WOLZA       Wolski  Zajazd  Zbyszek Piestrzeniewicz   
828      WOLZA       Wolski  Zajazd  Zbyszek Piestrzeniewicz   
829      WOLZA       Wolski  Zajazd  Zbyszek Piestrzeniewicz   

             contacttitle          address      city region postalcode  \
0    Sales Representative    Obere Str. 57    Berlin   None      12209   
1    Sales Representative    Obere Str. 57    Berlin   None      12209   
2    Sales Representative    Obere Str. 57    Berlin   None      12209   
3    Sales Representative    Obere Str. 57    Berlin   None      12209   
4    Sales Representative    Obere Str. 57    Berlin   None      12209   
..                    ...              ...       ...    ...        ...   
825                 Owner  ul. Filtrowa 68  Warszawa   None     01-012   
826                 Owner  ul. Filtrowa 68  Warszawa   None     01-012   
827                 Owner  ul. Filtrowa 68  Warszawa   None     01-012   
828                 Owner  ul. Filtrowa 68  Warszawa   None     01-012   
829                 Owner  ul. Filtrowa 68  Warszawa   None     01-012   

     country          phone            fax  orderid  employeeid   orderdate  \
0    Germany    030-0074321    030-0076545    10643           6  1997-08-25   
1    Germany    030-0074321    030-0076545    10692           4  1997-10-03   
2    Germany    030-0074321    030-0076545    10702           4  1997-10-13   
3    Germany    030-0074321    030-0076545    10835           1  1998-01-15   
4    Germany    030-0074321    030-0076545    10952           1  1998-03-16   
..       ...            ...            ...      ...         ...         ...   
825   Poland  (26) 642-7012  (26) 642-7012    10792           1  1997-12-23   
826   Poland  (26) 642-7012  (26) 642-7012    10870           5  1998-02-04   
827   Poland  (26) 642-7012  (26) 642-7012    10906           4  1998-02-25   
828   Poland  (26) 642-7012  (26) 642-7012    10998           8  1998-04-03   
829   Poland  (26) 642-7012  (26) 642-7012    11044           4  1998-04-23   

    requireddate shippeddate  shipvia  freight              shipname  \
0     1997-09-22  1997-09-02        1    29.46   Alfreds Futterkiste   
1     1997-10-31  1997-10-13        2    61.02  Alfred's Futterkiste   
2     1997-11-24  1997-10-21        1    23.94  Alfred's Futterkiste   
3     1998-02-12  1998-01-21        3    69.53  Alfred's Futterkiste   
4     1998-04-27  1998-03-24        1    40.42  Alfred's Futterkiste   
..           ...         ...      ...      ...                   ...   
825   1998-01-20  1997-12-31        3    23.79         Wolski Zajazd   
826   1998-03-04  1998-02-13        3    12.04         Wolski Zajazd   
827   1998-03-11  1998-03-03        3    26.29         Wolski Zajazd   
828   1998-04-17  1998-04-17        2    20.31         Wolski Zajazd   
829   1998-05-21  1998-05-01        1     8.72         Wolski Zajazd   

         shipaddress  shipcity shipregion shippostalcode shipcountry  
0      Obere Str. 57    Berlin       None          12209     Germany  
1      Obere Str. 57    Berlin       None          12209     Germany  
2      Obere Str. 57    Berlin       None          12209     Germany  
3      Obere Str. 57    Berlin       None          12209     Germany  
4      Obere Str. 57    Berlin       None          12209     Germany  
..               ...       ...        ...            ...         ...  
825  ul. Filtrowa 68  Warszawa     

In [50]:
students = pd.DataFrame(dict(sid=[1,2,3], sname=list('abc'), age=[18]*3))
scores = pd.DataFrame(dict(sid=[1,2,2,4], scores=[10,20,20,40], age=[18]*4 ))

students
scores

pd.merge(scores, students, how='inner', on='sid')
pd.merge(scores, students, how='left', on='sid')
pd.merge(scores, students, how='outer', on='sid')
# pd.merge(scores, students, how='cross')

sid sname  age
0    1     a   18
1    2     b   18
2    3     c   18

sid  scores  age
0    1      10   18
1    2      20   18
2    2      20   18
3    4      40   18

sid  scores  age_x sname  age_y
0    1      10     18     a     18
1    2      20     18     b     18
2    2      20     18     b     18

sid  scores  age_x sname  age_y
0    1      10     18     a   18.0
1    2      20     18     b   18.0
2    2      20     18     b   18.0
3    4      40     18   NaN    NaN

sid  scores  age_x sname  age_y
0    1    10.0   18.0     a   18.0
1    2    20.0   18.0     b   18.0
2    2    20.0   18.0     b   18.0
3    4    40.0   18.0   NaN    NaN
4    3     NaN    NaN     c   18.0

## Movies

In [51]:
movie_cols = ['movie_id', 'title']
movies = pd.read_table('data/u.item', sep='|', header=None, names=movie_cols, usecols=[0, 1])
movies.head()

movie_id              title
0         1   Toy Story (1995)
1         2   GoldenEye (1995)
2         3  Four Rooms (1995)
3         4  Get Shorty (1995)
4         5     Copycat (1995)

In [52]:
movies.shape

(1682, 2)

In [53]:
movies.movie_id.nunique()

1682

## Ratings

In [54]:
rating_cols = ['user_id', 'movie_id', 'rating', 'timestamp']
ratings = pd.read_table('data/u.data', sep='\t', header=None, names=rating_cols)
ratings.head()

user_id  movie_id  rating  timestamp
0      196       242       3  881250949
1      186       302       3  891717742
2       22       377       1  878887116
3      244        51       2  880606923
4      166       346       1  886397596

In [55]:
ratings.shape

(100000, 4)

In [57]:
ratings.movie_id.nunique()

1682

In [60]:
movies.head()
ratings.head()

movies.merge(ratings, on='movie_id')
ratings.merge(movies, on='movie_id')

movie_id              title
0         1   Toy Story (1995)
1         2   GoldenEye (1995)
2         3  Four Rooms (1995)
3         4  Get Shorty (1995)
4         5     Copycat (1995)

user_id  movie_id  rating  timestamp
0      196       242       3  881250949
1      186       302       3  891717742
2       22       377       1  878887116
3      244        51       2  880606923
4      166       346       1  886397596

movie_id                                      title  user_id  rating  \
0             1                           Toy Story (1995)      308       4   
1             1                           Toy Story (1995)      287       5   
2             1                           Toy Story (1995)      148       4   
3             1                           Toy Story (1995)      280       4   
4             1                           Toy Story (1995)       66       3   
...         ...                                        ...      ...     ...   
99995      1678                          Mat' i syn (1997)      863       1   
99996      1679                           B. Monkey (1998)      863       3   
99997      1680                       Sliding Doors (1998)      863       2   
99998      1681                        You So Crazy (1994)      896       3   
99999      1682  Scream of Stone (Schrei aus Stein) (1991)      916       3   

       timestamp  
0      887736532  
1      875334088  
2      877019411  
3      891700426  
4      883601324  
...          ...  
99995  889289570  
99996  889289491  
99997  889289570  
99998  887160722  
99999  880845755  

[100000 rows x 5 columns]

user_id  movie_id  rating  timestamp  \
0          196       242       3  881250949   
1           63       242       3  875747190   
2          226       242       5  883888671   
3          154       242       3  879138235   
4          306       242       5  876503793   
...        ...       ...     ...        ...   
99995      840      1674       4  891211682   
99996      655      1640       3  888474646   
99997      655      1637       3  888984255   
99998      655      1630       3  887428735   
99999      655      1641       3  887427810   

                                                   title  
0                                           Kolya (1996)  
1                                           Kolya (1996)  
2                                           Kolya (1996)  
3                                           Kolya (1996)  
4                                           Kolya (1996)  
...                                                  ...  
99995                                  Mamma Roma (1962)  
99996                             Eighth Day, The (1996)  
99997                                  Girls Town (1996)  
99998  Silence of the Palace, The (Saimt el Qusur) (1...  
99999                                    Dadetown (1995)  

[100000 rows x 5 columns]

In [61]:
ratings.loc[ratings.movie_id == 1, :].head()

user_id  movie_id  rating  timestamp
24        308         1       4  887736532
454       287         1       5  875334088
957       148         1       4  877019411
971       280         1       4  891700426
1324       66         1       3  883601324

## Merging Movies and Ratings

Let's pretend that you want to examine the ratings DataFrame, but you want to know the **title** of each movie rather than its **movie_id**. The best way to accomplish this objective is by "joining" (or "merging") the DataFrames using the Pandas `merge` function:

In [62]:
movies.columns

Index(['movie_id', 'title'], dtype='object')

In [63]:
ratings.columns

Index(['user_id', 'movie_id', 'rating', 'timestamp'], dtype='object')

In [64]:
movie_ratings = pd.merge(movies, ratings)
movie_ratings.columns

Index(['movie_id', 'title', 'user_id', 'rating', 'timestamp'], dtype='object')

In [69]:
len(movie_ratings.query("movie_id==1"))
len(ratings.query("movie_id==1"))



452

452

In [70]:
movie_ratings.head()

movie_id             title  user_id  rating  timestamp
0         1  Toy Story (1995)      308       4  887736532
1         1  Toy Story (1995)      287       5  875334088
2         1  Toy Story (1995)      148       4  877019411
3         1  Toy Story (1995)      280       4  891700426
4         1  Toy Story (1995)       66       3  883601324

In [71]:
movie_ratings.shape

(100000, 5)

Here's what just happened:

- Pandas noticed that movies and ratings had one column in common, namely **movie_id**. This is the "key" on which the DataFrames will be joined.
- The first **movie_id** in movies is 1. Thus, Pandas looked through every row in the ratings DataFrame, searching for a movie_id of 1. Every time it found such a row, it recorded the **user_id**, **rating**, and **timestamp** listed in that row. In this case, it found 452 matching rows.
- The second **movie_id** in movies is 2. Again, Pandas did a search of ratings and found 131 matching rows.
- This process was repeated for all of the remaining rows in movies.

At the end of the process, the movie_ratings DataFrame is created, which contains the two columns from movies (**movie_id** and **title**) and the three other colums from ratings (**user_id**, **rating**, and **timestamp**).

- **movie_id** 1 and its **title** are listed 452 times, next to the **user_id**, **rating**, and **timestamp** for each of the 452 matching ratings.
- **movie_id** 2 and its **title** are listed 131 times, next to the **user_id**, **rating**, and **timestamp** for each of the 131 matching ratings.
- And so on, for every movie in the dataset.

In [73]:
print(movies.shape)
print(ratings.shape)
print(movie_ratings.shape)

(1682, 2)
(100000, 4)
(100000, 5)


Notice the shapes of the three DataFrames:

- There are 1682 rows in the movies DataFrame.
- There are 100000 rows in the ratings DataFrame.
- The `merge` function resulted in a movie_ratings DataFrame with 100000 rows, because every row from ratings matched a row from movies.
- The movie_ratings DataFrame has 5 columns, namely the 2 columns from movies, plus the 4 columns from ratings, minus the 1 column in common.

By default, the `merge` function joins the DataFrames using all column names that are in common (**movie_id**, in this case). The [documentation](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.merge.html) explains how you can override this behavior.

# Part 3: What if...?

## What if the columns you want to join on don't have the same name?

In [74]:
movies.columns = ['m_id', 'title']
movies.columns

Index(['m_id', 'title'], dtype='object')

In [75]:
ratings.columns

Index(['user_id', 'movie_id', 'rating', 'timestamp'], dtype='object')

In [76]:
pd.merge(movies, ratings, left_on='m_id', right_on='movie_id').head()

m_id             title  user_id  movie_id  rating  timestamp
0     1  Toy Story (1995)      308         1       4  887736532
1     1  Toy Story (1995)      287         1       5  875334088
2     1  Toy Story (1995)      148         1       4  877019411
3     1  Toy Story (1995)      280         1       4  891700426
4     1  Toy Story (1995)       66         1       3  883601324

## What if you want to join on one index?

In [81]:
movies = movies.set_index('m_id')
movies.head()

title
m_id                   
1      Toy Story (1995)
2      GoldenEye (1995)
3     Four Rooms (1995)
4     Get Shorty (1995)
5        Copycat (1995)

In [86]:
print(ratings.query("movie_id==1").head())
print(pd.merge(movies, ratings, left_index=True, right_on='movie_id').head())
#       user_id  movie_id  rating  timestamp
# 24        308         1       4  887736532
# 454       287         1       5  875334088
# 957       148         1       4  877019411
# 971       280         1       4  891700426
# 1324       66         1       3  883601324
#                  title  user_id  movie_id  rating  timestamp
# 24    Toy Story (1995)      308         1       4  887736532
# 454   Toy Story (1995)      287         1       5  875334088
# 957   Toy Story (1995)      148         1       4  877019411
# 971   Toy Story (1995)      280         1       4  891700426
# 1324  Toy Story (1995)       66         1       3  883601324
# print(ratings.query("movie_id==1").head())
print(pd.merge(movies, ratings, left_index=True, right_on='movie_id').head())



      user_id  movie_id  rating  timestamp
24        308         1       4  887736532
454       287         1       5  875334088
957       148         1       4  877019411
971       280         1       4  891700426
1324       66         1       3  883601324
                 title  user_id  movie_id  rating  timestamp
24    Toy Story (1995)      308         1       4  887736532
454   Toy Story (1995)      287         1       5  875334088
957   Toy Story (1995)      148         1       4  877019411
971   Toy Story (1995)      280         1       4  891700426
1324  Toy Story (1995)       66         1       3  883601324


title  user_id  movie_id  rating  timestamp
24    Toy Story (1995)      308         1       4  887736532
454   Toy Story (1995)      287         1       5  875334088
957   Toy Story (1995)      148         1       4  877019411
971   Toy Story (1995)      280         1       4  891700426
1324  Toy Story (1995)       66         1       3  883601324

## What if you want to join on two indexes?

In [20]:
ratings = ratings.set_index('movie_id')
ratings.head()

user_id  rating  timestamp
movie_id                            
242           196       3  881250949
302           186       3  891717742
377            22       1  878887116
51            244       2  880606923
346           166       1  886397596

In [21]:
pd.merge(movies, ratings, left_index=True, right_index=True).head()

title  user_id  rating  timestamp
1  Toy Story (1995)      308       4  887736532
1  Toy Story (1995)      287       5  875334088
1  Toy Story (1995)      148       4  877019411
1  Toy Story (1995)      280       4  891700426
1  Toy Story (1995)       66       3  883601324

# Part 4: Four Types of Joins

There are actually four types of joins supported by the Pandas `merge` function. Here's how they are described by the documentation:

- **inner:** use intersection of keys from both frames, similar to a SQL inner join; preserve the order of the left keys
- **outer:** use union of keys from both frames, similar to a SQL full outer join; sort keys lexicographically
- **left:** use only keys from left frame, similar to a SQL left outer join; preserve key order
- **right:** use only keys from right frame, similar to a SQL right outer join; preserve key order

The default is the "inner join", which was used when creating the movie_ratings DataFrame.

It's easiest to understand the different types by looking at some simple examples:

## Example DataFrames A and B

In [22]:
A = pd.DataFrame({'color': ['green', 'yellow', 'red'], 'num':[1, 2, 3]})
A

color  num
0   green    1
1  yellow    2
2     red    3

In [23]:
B = pd.DataFrame({'color': ['green', 'yellow', 'pink'], 'size':['S', 'M', 'L']})
B

color size
0   green    S
1  yellow    M
2    pink    L

## Inner join

Only include observations found in both A and B:

In [24]:
pd.merge(A, B, how='inner')

color  num size
0   green    1    S
1  yellow    2    M

## Outer join

Include observations found in either A or B:

In [25]:
pd.merge(A, B, how='outer')

color  num size
0   green  1.0    S
1  yellow  2.0    M
2     red  3.0  NaN
3    pink  NaN    L

## Left join

Include all observations found in A:

In [26]:
pd.merge(A, B, how='left')

color  num size
0   green    1    S
1  yellow    2    M
2     red    3  NaN

## Right join

Include all observations found in B:

In [27]:
pd.merge(A, B, how='right')

color  num size
0   green  1.0    S
1  yellow  2.0    M
2    pink  NaN    L